File: app/src/test/java/org/thoughtcrime/securesms/payments/MobileCoinPublicAddressProfileUtilTest.java
package org.thoughtcrime.securesms.payments;

import org.junit.Before;
import org.junit.Test;
import org.signal.libsignal.protocol.IdentityKey;
import org.signal.libsignal.protocol.IdentityKeyPair;
import org.thoughtcrime.securesms.crypto.IdentityKeyUtil;
import org.thoughtcrime.securesms.util.Util;
import org.whispersystems.signalservice.internal.push.PaymentAddress;

import okio.ByteString;

import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.Assert.assertArrayEquals;
import static org.whispersystems.signalservice.test.LibSignalLibraryUtil.assumeLibSignalSupportedOnOS;

public final class MobileCoinPublicAddressProfileUtilTest {

  @Before
  public void ensureNativeSupported() {
    assumeLibSignalSupportedOnOS();
  }

  @Test
  public void can_verify_an_address() throws PaymentsAddressException {
    IdentityKeyPair identityKeyPair      = IdentityKeyUtil.generateIdentityKeyPair();
    byte[]          address              = Util.getSecretBytes(100);
    PaymentAddress  signedPaymentAddress = MobileCoinPublicAddressProfileUtil.signPaymentsAddress(address, identityKeyPair);

    byte[] paymentsAddress = MobileCoinPublicAddressProfileUtil.verifyPaymentsAddress(signedPaymentAddress, identityKeyPair.getPublicKey());

    assertArrayEquals(address, paymentsAddress);
  }

  @Test
  public void can_not_verify_an_address_with_the_wrong_key() {
    IdentityKeyPair identityKeyPair      = IdentityKeyUtil.generateIdentityKeyPair();
    IdentityKey     wrongPublicKey       = IdentityKeyUtil.generateIdentityKeyPair().getPublicKey();
    byte[]          address              = Util.getSecretBytes(100);
    PaymentAddress  signedPaymentAddress = MobileCoinPublicAddressProfileUtil.signPaymentsAddress(address, identityKeyPair);

    assertThatThrownBy(() -> MobileCoinPublicAddressProfileUtil.verifyPaymentsAddress(signedPaymentAddress, wrongPublicKey))
        .isInstanceOf(PaymentsAddressException.class)
        .hasMessage("Invalid MobileCoin address signature on payments address proto");
  }

  @Test
  public void can_not_verify_a_tampered_signature() {
    IdentityKeyPair identityKeyPair      = IdentityKeyUtil.generateIdentityKeyPair();
    byte[]          address              = Util.getSecretBytes(100);
    PaymentAddress  signedPaymentAddress = MobileCoinPublicAddressProfileUtil.signPaymentsAddress(address, identityKeyPair);

    byte[] signature = signedPaymentAddress.mobileCoinAddress.signature.toByteArray();
    signature[0] = (byte) (signature[0] ^ 0x01);
    PaymentAddress tamperedSignature = signedPaymentAddress.newBuilder()
                                                           .mobileCoinAddress(signedPaymentAddress.mobileCoinAddress
                                                                                  .newBuilder()
                                                                                  .signature(ByteString.of(signature))
                                                                                  .build())
                                                           .build();

    assertThatThrownBy(() -> MobileCoinPublicAddressProfileUtil.verifyPaymentsAddress(tamperedSignature, identityKeyPair.getPublicKey()))
        .isInstanceOf(PaymentsAddressException.class)
        .hasMessage("Invalid MobileCoin address signature on payments address proto");
  }

  @Test
  public void can_not_verify_a_tampered_address() {
    IdentityKeyPair identityKeyPair      = IdentityKeyUtil.generateIdentityKeyPair();
    byte[]          addressBytes         = Util.getSecretBytes(100);
    PaymentAddress  signedPaymentAddress = MobileCoinPublicAddressProfileUtil.signPaymentsAddress(addressBytes, identityKeyPair);

    byte[] address = signedPaymentAddress.mobileCoinAddress.address.toByteArray();
    address[0] = (byte) (address[0] ^ 0x01);
    PaymentAddress tamperedAddress = signedPaymentAddress.newBuilder()
                                                         .mobileCoinAddress(signedPaymentAddress.mobileCoinAddress
                                                                                .newBuilder()
                                                                                .address(ByteString.of(address))
                                                                                .build())
                                                         .build();

    assertThatThrownBy(() -> MobileCoinPublicAddressProfileUtil.verifyPaymentsAddress(tamperedAddress, identityKeyPair.getPublicKey()))
        .isInstanceOf(PaymentsAddressException.class)
        .hasMessage("Invalid MobileCoin address signature on payments address proto");
  }

  @Test
  public void can_not_verify_a_missing_signature() {
    IdentityKeyPair identityKeyPair      = IdentityKeyUtil.generateIdentityKeyPair();
    byte[]          address              = Util.getSecretBytes(100);
    PaymentAddress  signedPaymentAddress = MobileCoinPublicAddressProfileUtil.signPaymentsAddress(address, identityKeyPair);

    PaymentAddress removedSignature = signedPaymentAddress.newBuilder()
                                                          .mobileCoinAddress(signedPaymentAddress.mobileCoinAddress
                                                                                 .newBuilder()
                                                                                 .signature(null)
                                                                                 .build())
                                                          .build();

    assertThatThrownBy(() -> MobileCoinPublicAddressProfileUtil.verifyPaymentsAddress(removedSignature, identityKeyPair.getPublicKey()))
        .isInstanceOf(PaymentsAddressException.class)
        .hasMessage("Invalid MobileCoin address signature on payments address proto");
  }

  @Test
  public void can_not_verify_a_missing_address() {
    IdentityKeyPair identityKeyPair      = IdentityKeyUtil.generateIdentityKeyPair();
    byte[]          address              = Util.getSecretBytes(100);
    PaymentAddress  signedPaymentAddress = MobileCoinPublicAddressProfileUtil.signPaymentsAddress(address, identityKeyPair);

    PaymentAddress removedAddress = signedPaymentAddress.newBuilder()
                                                        .mobileCoinAddress(signedPaymentAddress.mobileCoinAddress
                                                                               .newBuilder()
                                                                               .address(null)
                                                                               .build())
                                                        .build();

    assertThatThrownBy(() -> MobileCoinPublicAddressProfileUtil.verifyPaymentsAddress(removedAddress, identityKeyPair.getPublicKey()))
        .isInstanceOf(PaymentsAddressException.class)
        .hasMessage("Invalid MobileCoin address signature on payments address proto");
  }
}

File: app/src/test/java/org/thoughtcrime/securesms/payments/MobileCoinPublicAddressProfileUtilTest.kt
package org.thoughtcrime.securesms.payments

import assertk.assertFailure
import assertk.assertThat
import assertk.assertions.hasMessage
import assertk.assertions.isEqualTo
import assertk.assertions.isInstanceOf
import okio.ByteString
import org.junit.Before
import org.junit.Test
import org.thoughtcrime.securesms.crypto.IdentityKeyUtil
import org.thoughtcrime.securesms.util.Util
import org.whispersystems.signalservice.test.LibSignalLibraryUtil

class MobileCoinPublicAddressProfileUtilTest {
  @Before
  fun ensureNativeSupported() {
    LibSignalLibraryUtil.assumeLibSignalSupportedOnOS()
  }

  @Test
  fun can_verify_an_address() {
    val identityKeyPair = IdentityKeyUtil.generateIdentityKeyPair()
    val address = Util.getSecretBytes(100)
    val signedPaymentAddress = MobileCoinPublicAddressProfileUtil.signPaymentsAddress(address, identityKeyPair)

    val paymentsAddress = MobileCoinPublicAddressProfileUtil.verifyPaymentsAddress(signedPaymentAddress, identityKeyPair.publicKey)

    assertThat(paymentsAddress).isEqualTo(address)
  }

  @Test
  fun can_not_verify_an_address_with_the_wrong_key() {
    val identityKeyPair = IdentityKeyUtil.generateIdentityKeyPair()
    val wrongPublicKey = IdentityKeyUtil.generateIdentityKeyPair().publicKey
    val address = Util.getSecretBytes(100)
    val signedPaymentAddress = MobileCoinPublicAddressProfileUtil.signPaymentsAddress(address, identityKeyPair)

    assertFailure { MobileCoinPublicAddressProfileUtil.verifyPaymentsAddress(signedPaymentAddress, wrongPublicKey) }
      .isInstanceOf<PaymentsAddressException>()
      .hasMessage("Invalid MobileCoin address signature on payments address proto")
  }

  @Test
  fun can_not_verify_a_tampered_signature() {
    val identityKeyPair = IdentityKeyUtil.generateIdentityKeyPair()
    val address = Util.getSecretBytes(100)
    val signedPaymentAddress = MobileCoinPublicAddressProfileUtil.signPaymentsAddress(address, identityKeyPair)
    val mobileCoinAddress = signedPaymentAddress.mobileCoin!!

    val signature = mobileCoinAddress.signature!!.toByteArray()
    signature[0] = (signature[0].toInt() xor 0x01).toByte()
    val tamperedSignature = signedPaymentAddress.newBuilder()
      .mobileCoin(
        mobileCoinAddress
          .newBuilder()
          .signature(ByteString.of(*signature))
          .build()
      )
      .build()

    assertFailure { MobileCoinPublicAddressProfileUtil.verifyPaymentsAddress(tamperedSignature, identityKeyPair.publicKey) }
      .isInstanceOf<PaymentsAddressException>()
      .hasMessage("Invalid MobileCoin address signature on payments address proto")
  }

  @Test
  fun can_not_verify_a_tampered_address() {
    val identityKeyPair = IdentityKeyUtil.generateIdentityKeyPair()
    val addressBytes = Util.getSecretBytes(100)
    val signedPaymentAddress = MobileCoinPublicAddressProfileUtil.signPaymentsAddress(addressBytes, identityKeyPair)
    val mobileCoinAddress = signedPaymentAddress.mobileCoin!!

    val address = mobileCoinAddress.publicAddress!!.toByteArray()
    address[0] = (address[0].toInt() xor 0x01).toByte()
    val tamperedAddress = signedPaymentAddress.newBuilder()
      .mobileCoin(
        mobileCoinAddress
          .newBuilder()
          .publicAddress(ByteString.of(*address))
          .build()
      )
      .build()

    assertFailure { MobileCoinPublicAddressProfileUtil.verifyPaymentsAddress(tamperedAddress, identityKeyPair.publicKey) }
      .isInstanceOf<PaymentsAddressException>()
      .hasMessage("Invalid MobileCoin address signature on payments address proto")
  }

  @Test
  fun can_not_verify_a_missing_signature() {
    val identityKeyPair = IdentityKeyUtil.generateIdentityKeyPair()
    val address = Util.getSecretBytes(100)
    val signedPaymentAddress = MobileCoinPublicAddressProfileUtil.signPaymentsAddress(address, identityKeyPair)

    val removedSignature = signedPaymentAddress.newBuilder()
      .mobileCoin(
        signedPaymentAddress.mobileCoin!!
          .newBuilder()
          .signature(null)
          .build()
      )
      .build()

    assertFailure { MobileCoinPublicAddressProfileUtil.verifyPaymentsAddress(removedSignature, identityKeyPair.publicKey) }
      .isInstanceOf<PaymentsAddressException>()
      .hasMessage("Invalid MobileCoin address signature on payments address proto")
  }

  @Test
  fun can_not_verify_a_missing_address() {
    val identityKeyPair = IdentityKeyUtil.generateIdentityKeyPair()
    val address = Util.getSecretBytes(100)
    val signedPaymentAddress = MobileCoinPublicAddressProfileUtil.signPaymentsAddress(address, identityKeyPair)

    val removedAddress = signedPaymentAddress.newBuilder()
      .mobileCoin(
        signedPaymentAddress.mobileCoin!!
          .newBuilder()
          .publicAddress(null)
          .build()
      )
      .build()

    assertFailure { MobileCoinPublicAddressProfileUtil.verifyPaymentsAddress(removedAddress, identityKeyPair.publicKey) }
      .isInstanceOf<PaymentsAddressException>()
      .hasMessage("Invalid MobileCoin address signature on payments address proto")
  }
}
